File: BaekjoonOnlineJudge/1197/1197.py
import sys, heapq

input = sys.stdin.readline


def find(target):
    if target == disjoint_set[target]:
        return target
    
    disjoint_set[target] = find(disjoint_set[target])
    return disjoint_set[target]


def union(a, b):
    find_a = find(a)
    find_b = find(b)

    if find_a == find_b:
        return False
    elif find_a < find_b:
        disjoint_set[find_b] = find_a
    else:
        disjoint_set[find_a] = find_b

    return True

def kruskal():
    global answer

    while hq:
        weight, a, b = heapq.heappop(hq)
    
        if union(a, b):
            answer += weight

V, E = map(int, input().split())
disjoint_set = list(range(V+1))
answer = 0

hq = []

for _ in range(E):
    A, B, C = map(int, input().split())

    heapq.heappush(hq, (C, A, B))

kruskal()

print(answer)

File: BaekjoonOnlineJudge/1197/1197.java
import java.io.*;
import java.util.*;

class Edge {

    int weight;
    int from;
    int to;

    public Edge(int weight, int from, int to) {
        this.weight = weight;
        this.from = from;
        this.to = to;
    }
}

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;

    private static int v, e;
    private static int[] disjointSet;
    private static PriorityQueue<Edge> edgePq = new PriorityQueue<>(
        (e1, e2) -> e1.weight - e2.weight);
    private static int answer = 0;

    private static int find(int target) {
        if (disjointSet[target] == target) {
            return target;
        }

        return disjointSet[target] = find(disjointSet[target]);
    }

    private static boolean union(int e1, int e2) {
        int findE1 = find(e1);
        int findE2 = find(e2);

        if (findE1 == findE2) {
            return false;
        }

        if (findE1 < findE2) {
            disjointSet[findE2] = findE1;
        } else {
            disjointSet[findE1] = findE2;
        }

        return true;
    }

    private static void kruskal() {
        while (!edgePq.isEmpty()) {
            Edge e = edgePq.poll();

            if (union(e.from, e.to)) {
                answer += e.weight;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        v = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());

        disjointSet = new int[v + 1];

        for (int i = 1; i < v + 1; i++) {
            disjointSet[i] = i;
        }

        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());

            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());

            edgePq.add(new Edge(weight, from, to));
        }

        kruskal();

        bw.write(String.valueOf(answer));

        bw.flush();
        bw.close();
        br.close();
    }
}